### PR TITLE
Add array path support for deletedIDFieldName in NODE_DELETE mutations

### PR DESCRIPTION
**Pull request by vangavroche**
for issue #1361 @wincent

### CODE REVIEW

**Comment by facebook-github-bot-1** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-8** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: src/traversal/__tests__/writeRelayUpdatePayload-test.js [hunk 1]
```diff
@@ -1078,6 +1078,196 @@ describe('writeRelayUpdatePayload()', () => {
       // connection metadata is merged into the queued store
       expect(store.getField(connectionID, 'count')).toBe(0);
     });
+
+    // test cases for array path deletedIDFieldName
```
Comment (wincent) [inline src/traversal/__tests__/writeRelayUpdatePayload-test.js]: This comment seems redundant. If you wanted to get rid of the repeated references to "array path deletedIDFieldName" below you could use a nested `describe`:

```
describe('with an array path `deletedIDFieldName`', () => {
  it('optimistically deletes comments, () => { /* ... */ });
  it('non-optimistically deletes comments, () => { /* ... */ });
});
```

File: src/traversal/__tests__/writeRelayUpdatePayload-test.js [hunk 2]
```diff
@@ -1261,6 +1451,8 @@ describe('writeRelayUpdatePayload()', () => {
     });
   });
 
+
+
```
Comment (wincent) [inline src/traversal/__tests__/writeRelayUpdatePayload-test.js]: Probably don't want this whitespace change.

File: src/traversal/writeRelayUpdatePayload.js [hunk 3]
```diff
@@ -187,6 +199,17 @@ function handleMerge(
     if (!Object.prototype.hasOwnProperty.call(payload, fieldName)) {
       continue;
     }
+
+    // if fieldName is part of deletedIDFieldName in some NODE_DELETE config
+    if (configs.some((config) => {
+      return config.type === RelayMutationType.NODE_DELETE &&
+        Array.isArray(config.deletedIDFieldName) &&
+        config.deletedIDFieldName.length > 0 &&
```
Comment (wincent) [inline src/traversal/writeRelayUpdatePayload.js]: By convention, we usually express this type of thing as just:

```
config.deletedIDFieldName.length
```

File: src/traversal/writeRelayUpdatePayload.js [hunk 4]
```diff
@@ -187,6 +199,17 @@ function handleMerge(
...
+        config.deletedIDFieldName[0] === fieldName;
+    })) {
```
Comment (wincent) [inline src/traversal/writeRelayUpdatePayload.js]: Style nit (non-blocking feedback): would probably make this:

```
if (configs.some(config => (
  config.type === ... &&
  Array.isArray(...) &&
  config ... &&
  config ...
)) {
  continue;
}
```

(ie. no parens around single arrow function param, and no need for explicit `return`)

**Comment by wincent** [comment]
Thanks for this @vangavroche. I've left some superficial nit comments, but will take a proper look tomorrow. I have a diff to fix the unrelated Travis failures, which are due to something bad that I merged earlier today.

File: src/traversal/writeRelayUpdatePayload.js (hunk 3 above)
Comment (vangavroche) [inline src/traversal/writeRelayUpdatePayload.js]: Isn't config.deletedIDFieldName.length > 0 more readable >_< ?

Comment (wincent) [inline src/traversal/writeRelayUpdatePayload.js]: Many people would think so. But many others would think the opposite. In other words, it is a bikeshed topic.

One thing that pretty much everyone agrees on, however, is that there is much value in being consistent, picking a pattern and sticking to it, and the pattern we've tried to follow is the one I described.

Comment (vangavroche) [inline src/traversal/writeRelayUpdatePayload.js]: Okay. I understand.

**Comment by wincent** [comment]
@facebook-github-bot import

**Comment by wincent** [comment]
This is hopelessly stale now. It's a shame the import did not work. If you still have interest in this, would you mind submitting a new PR? We're unlikely to do any work against the legacy APIs at this time, now that Relay Modern is out.

**Pull request closed without merging**